[김민찬] Sprint1

itscold96 opened the pull request:
배포 주소: [panda-market-itscold96.netlify.app](https://panda-market-itscold96.netlify.app/)

## 요구사항

### 기본
- [x] 랜딩 페이지의 url path는 루트(‘/’)로 설정합니다.
- [x] title은 “판다마켓”로 설정합니다.
- [x] 화면의 너비가 1920px 이상이면 하늘색 배경색은 너비를 꽉 채우도록 채워지고, 내부 요소들의 위치는 고정되고, 여백만 커지도록 합니다.
- [x] 화면의 너비가 1920px 보다 작아질 때, “판다마켓” 로고의 왼쪽 여백 200px“로그인" 버튼의 오른쪽 여백 200px이 유지되고, 화면의 너비가 작아질수록 두 요소간 거리가 가까워지도록 설정합니다.
- [x] 클릭으로 기능이 동작해야 하는 경우, 사용자가 클릭할 수 있는 요소임을 알 수 있도록 CSS 속성 cursor: pointer 로 설정합니다.
- [x] “판다마켓” 클릭 시 루트 페이지(‘/’)로 이동합니다.
- [x] '로그인'버튼 클릭 시 로그인 페이지(‘/login’)로 이동합니다 (빈 페이지)
- [x] “구경하러가기”버튼 클릭 시(’/items’)로 이동합니다.(빈 페이지)
- [x] “Privacy Policy”, “FAQ”는 클릭 시 각각 Privacy 페이지(‘/privacy’), FAQ 페이지(‘/faq’)로 이동합니다.(모두 빈 페이지)
- [x] 페이스북, 트위터, 유튜브, 인스타그램 아이콘을 클릭 시 각각의 홈페이지로 새로운 창이 열리면서 이동합니다.

### 심화

- [x] palette에 있는 color값들을 css 변수로 등록하고 사용해 주세요.
- [x] 사용자의 브라우저가 크고 작아짐에 따라 페이지의 요소간 간격, 요소의 크기, font-size 등 모든 크기와 관련된 값이 크고 작아지도록 설정해 보세요.(설정값은 자유입니다)

## 주요 변경사항
- 반응형 요소 사이즈 변화 구현
- css 모듈화

## 스크린샷
<img width="2047" alt="스크린샷 2024-05-24 오후 10 51 56" src="https://github.com/itscold96/Next-NaverMap/assets/98478661/1e10f051-6d65-4d3c-bc9d-021827ea3e3e">

![스크린샷 2024-05-24 오후 10 44 15](https://github.com/itscold96/Next-NaverMap/assets/98478661/743bb6ab-438e-40f9-aa75-e335feb0ae2c)

## 멘토에게
- 부족한 부분이 많을 겁니다. 가감없이 꼭 알려주시기 바랍니다.

Commit:
김민찬 첫번째 sprint mission 제출

Taero-Kim commented on index.html:
```diff
@@ -0,0 +1,146 @@
+          alt="앱 로고와 이름 이미지"
+        />
+      </a>
+      <button class="header-login" onclick="moveLocation('/login')">
```
p4;
'/login'이나 '/item' 같은 서비스 path를 상수로 관리하면 휴먼 에러를 더 줄일 수 있을 것 같아요.
상수관련 디렉터리를 하나 생성하여 js파일에 아래와 같이 관리하는 것도 좋아보여요!

ex)
const LOGIN_PAGE_URL = '/login';
const ITEMS_PAGE_URL = '/items';

itscold96 replied:
상수를 제작하는 것에 대한 부분은 제가 미처 생각지 못한 부분이고,
당연히 위와 같이 변경될 여지가 적고, 여러 곳에서 사용 되어 휴먼 에러의 여지가 있는 부분은
상수화 하는 것이 좋다고 생각합니다.

다만, 
이게 리액트 환경이라 JSX를 사용할 수 있다면 템플릿 리터럴로 작성하면 되기에 위와 같은 작업은 매우 간단할 것입니다.
하지만 현재 작업 환경이 리액트가 아닌 단순 HTML 파일에서 script로 HTML을 제어해야 하는 상황입니다.

이와 같은 상황에서, js 파일에 선언한 변수를 HTML 내부 태그의 속성값으로 가져오기 위해서는
getElemnetBy 류 메서드를 사용해서 script로 일일이 해당 태그를 찾아 속성을 집어넣어줘야 합니다.
 
저도 리액트를 오랫동안 써왔기에 JSX가 너무 몸에 베여있어서 간단한 작업일 것이라 생각했는데,
생각 외로 html 태그를 하나하나 직접 조작하는 것은 번거롭더군요.

위와 같은 작업을 하는 것이 제 기준에서 어려운 일은 아니나,
해당 스프린트 미션에서 요구하는 조건 이상의 것을 구현하고 있다는 생각이 들었습니다.

상수화는 매우 필요한 작업이지만,
현재 환경과 조건에서 상수화라는 작은 작업하기 위해,
코드잇 커리큘럼에서 배우지도 않은 getElemnetBy 메서드까지 써가면서
상수를 쓰는 것은 조금 힘든 부분이라 생각합니다.

이 부분에 대해서 어떻게 생각하시는지 알려주시면 감사하겠습니다.

ps. JSX로 얼마나 편하게 작업하고 있었는지 새삼 느꼈습니다..ㅎㅎ

Taero-Kim commented on index.html:
```diff
@@ -0,0 +1,146 @@
+        <a href="/faq">FAQ</a>
+      </div>
+      <div class="home-footer-sns">
+        <a href="https://www.facebook.com" target="_blank">
```
p4;
푸터부분의 SNS 링크들도 마찬가지로 객체 상수로 관리하면
휴먼에러를 방지하고 경우에 따라, 다른 곳에서 사용하거나 유지보수하기 용이할 것 같아요!

```constants/sns.js
const SNS_LINK_INFO = {
  FACEBOOK: {
    URL: 'https://www.facebook.com',
    IMAGE_SRC: './assets/...',
    ALT: '페이스북 로고'
  },
  ...
}
```

itscold96 replied:
상수를 제작하는 것에 대한 부분은 제가 미처 생각지 못한 부분이고,
당연히 위와 같이 변경될 여지가 적고, 여러 곳에서 사용 되어 휴먼 에러의 여지가 있는 부분은
상수화 하는 것이 좋다고 생각합니다.

다만,
이게 리액트 환경이라 JSX를 사용할 수 있다면
상수 파일 import 후 템플릿 리터럴로 JSX 속성값에 직접 상수를 기입하면 되기에 위와 같은 작업은 매우 간단할 것입니다.
하지만 현재 작업 환경이 리액트가 아닌 단순 HTML 파일에서 script로 HTML을 제어해야 하는 상황입니다.

이와 같은 상황에서, js 파일에 선언한 변수를 HTML 내부 태그의 속성값으로 가져오기 위해서는
getElemnetBy 류 메서드를 사용해서 script로 일일이 해당 태그를 찾아 속성을 집어넣어줘야 합니다.

저도 리액트를 오랫동안 써왔기에 JSX가 너무 몸에 베여있어서 간단한 작업일 것이라 생각했는데,
생각 외로 html 태그를 하나하나 직접 조작하는 것은 번거롭더군요.

위와 같은 작업을 하는 것이 제 기준에서 어려운 일은 아니나,
해당 스프린트 미션에서 요구하는 조건 이상의 것을 구현하고 있다는 생각이 들었습니다.

상수화는 매우 필요한 작업이지만,
현재 환경과 조건에서 상수화라는 작은 작업하기 위해,
코드잇 커리큘럼에서 배우지도 않은 getElemnetBy 메서드까지 써가면서
상수를 쓰는 것은 조금 주저되는 부분이라 생각합니다.

이 부분에 대해서 어떻게 생각하시는지 알려주시면 감사하겠습니다.

ps. JSX로 얼마나 편하게 작업하고 있었는지 새삼 느꼈습니다..ㅎㅎ

Taero-Kim replied:
맞는 말씀이십니다! 제가 JSX에 익숙해져서 jSX 기준으로 설명드린 것 같아요!
민찬님 말씀대로 오버코딩이고 오히려 더 코드가 지저분해질 것 같아요.
상수화라는 포인트만 잘 기억하고 그냥 무시하시고 넘어가셔도 좋을 것 같습니다👍

Taero-Kim commented on javascript/functions.js:
```diff
@@ -0,0 +1,3 @@
+moveLocation = (param) => {
+  return (location.href = `${param}`);
```
p2;
1. let, const 등의 변수 선언 방식을 명시하는 것이 좋을 것 같아요!
그렇지 않은 경우 함수가 암묵적으로 전역 스코프에 생성되어,
다른 스크립트 파일에서 중복으로 함수명이 사용될 경우 예상치 못하게 오버라이딩 되기도 하고,
브라우저 전역 객체(window)에 있는 함수를 오버라이딩 하기도 합니다.

물론, 이 경우 어차피 루트 Index.html에서 해당 js파일을 불러오기 때문에, 
사실상 const 키워드를 붙여도 전역 함수로 생성되는 것이긴 하지만,
그럼에도 불구하고 의도를 명확하게 하기 위해, 좋은 습관을 들이기 위해
꼭 변수 선은 키워드를 붙이는 것이 좋을 것 같아요.

2. moveLocation 함수는 단순히 location을 이동시키는 역할을 할 뿐입니다.
리턴하지 않아도 될 뿐더러, 리턴하는 것이 이 함수의 본질적인 의도에 어긋나는 것 같아요!

3. 불필요하게 템플릿 리터럴을 사용하지 않아도 될 것 같아요.
어차피 moveLocation에는 string 파라미터를 넘겨주고 있는데, 이걸 다시 템플릿 리터럴로 감싸지 않아도 될 것 같습니다.

기타 개인적인 사소한 의견) 파라미터 변수명을 직관적이게 수정하면 더 좋을 것 같아요!
- 동작에는 전혀 이상이 없고 함수명과 함수 내용만으로도 충분히 유추가 가능하지만,
param이라는 이름 대신 href등의 파라미터명을 사용해도 좋을 것 같아요!

```ts
const moveLocation = (href) => {
  location.href = href;
}
```

itscold96 replied:
반영 후 커밋하였습니다!

Taero-Kim commented on index.html:
```diff
@@ -0,0 +1,146 @@
+          버튼처럼 디자인이 필요한 경우에는 button 태그를 활용하여 onClick 함수를 구현하였습니다.
+          이유는 a 태그의 경우 기본적으로 display: inline 속성이기 때문에 css를 입히기 번거로우며,
+          해당 기능이 단순 페이지 이동이 아닌, 데이터 불러오기나 사용자 정보 확인과 같은 기능이 덧붙여질 가능성이 높기 때문에
+          onClick 함수의 사용이 더 적합하다고 생각하였습니다. -->
```
good!!

Taero-Kim commented:
작다면 작은 이 프로젝트에서 정말 엄청난 정성이 느껴지는 코드입니다.
미디어 쿼리와 rem을 사용해 브라우저 사이즈에 따른 반응형을 정말 잘 구현하셨습니다!

건드릴게 크게 없어보이는데 그 와중에 사소하게 몇가지 개선하면 좋을 부분에 대해서 코멘트 남겼습니다.

제가 남긴 리뷰에 대한 질문이나 기타 질문이 있으시면 언제든 편하게 말씀해주세요!

+) 리뷰를 남긴 코멘트 위의 p(1~5); 마크의 의미는 아래와 같습니다! 참고하면 좋을 것 같아요!

```
p1; 꼭 반영해주세요 (Request changes)
리뷰어는 PR의 내용이 서비스에 중대한 오류를 발생할 수 있는 가능성을 잠재하고 있는 등 중대한 코드 수정이 반드시 필요하다고 판단되는 경우, P1 태그를 통해 리뷰 요청자에게 수정을 요청합니다. 리뷰 요청자는 p1 태그에 대해 리뷰어의 요청을 반영하거나, 반영할 수 없는 합리적인 의견을 통해 리뷰어를 설득할 수 있어야 합니다.

p2; 적극적으로 고려해주세요 (Request changes)
작성자는 P2에 대해 수용하거나 만약 수용할 수 없는 상황이라면 적합한 의견을 들어 토론할 것을 권장합니다.

p3; 웬만하면 반영해 주세요 (Comment)
작성자는 P3에 대해 수용하거나 만약 수용할 수 없는 상황이라면 반영할 수 없는 이유를 들어 설명하거나 다음에 반영할 계획을 명시적으로(JIRA 티켓 등으로) 표현할 것을 권장합니다. Request changes 가 아닌 Comment 와 함께 사용됩니다.

p4; 반영해도 좋고 넘어가도 좋습니다 (Approve)
작성자는 P4에 대해서는 아무런 의견을 달지 않고 무시해도 괜찮습니다. 해당 의견을 반영하는 게 좋을지 고민해 보는 정도면 충분합니다.

p5; 그냥 사소한 의견입니다 (Approve)
작성자는 P5에 대해 아무런 의견을 달지 않고 무시해도 괜찮습니다.
```

itscold96 committed:
refactor(mentor): moveLocation 함수 수정

- const로 선언 후 매개변수 이름을 바꾸어 좀더 직관적으로 매개변수의 역할을 알 수 있도록 하였음
- 기존 단순 파일에서 util 폴더를 만들고 location.js로 이름을 바꾸어 폴더 구조와 파일 이름을 직관적으로 수정함

itscold96 committed:
fix(index.html): css 클래스명 변경 및 footer에서 a 태그들의 높이를 컨텐츠 크기에 맞춤

- 버튼 태그가 아님에도 .home-footer-buttons로 되어 있던 클래스명을
.home-footer-anchors로 수정

- footer에서 a 태그들의 높이가 박스가 가지는 최대 범위로 가지고 있던 것을
height: fit-content 속성을 추가하여 높이를 컨텐츠 크기에 맞춤

itscold96 committed:
refactor(mentor): 상수 제작

- 상수화된 변수를 생성
- 현재 진행하고 있는 프로젝트의 특성 상, HTML 파일 내에서는 JSX를 사용할 수 없기 때문에 템플릿 리터럴로 상수를 import 해오는 것이 불가능함
- 물론 getElementBy 류 메서드를 활용하여 속성과 내부 HTML을 생성하는 것은 가능한 부분이지만 그것은 이 스프린트 프로젝트의 활용 범위를 넘어서는 것이라 생각함
